aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/SetupTOTPCognitoActions.kt

```diff
@@ -0,0 +1,137 @@
+                        SignInTOTPSetupData(secret, response.session, eventType.totpSetupDetails.username)
+                    )
+                )
+            } ?: SetupTOTPEvent(
+                SetupTOTPEvent.EventType.ThrowAuthError(Exception("Software token setup failed"))
+            )
+        } catch (e: Exception) {
+            SetupTOTPEvent(
+                SetupTOTPEvent.EventType.ThrowAuthError(e)
+            )
+        }
+        logger.verbose("$id Sending event ${evt.type}")
+        dispatcher.send(evt)
+    }
+
+    override fun verifyChallengeAnswer(
+        eventType: SetupTOTPEvent.EventType.VerifyChallengeAnswer
+    ): Action =
+        Action<AuthEnvironment>("verifyChallengeAnswer") { id, dispatcher ->
+            logger.verbose("$id Starting execution")
+            val evt = try {
+                val response = cognitoAuthService.cognitoIdentityProviderClient?.verifySoftwareToken {
+                    userCode = eventType.answer
+                    this.session = eventType.session
+                    this.friendlyDeviceName = eventType.friendlyDeviceName
+                }
+
+                response?.let {
+                    when (it.status) {
+                        is VerifySoftwareTokenResponseType.Success -> {
+                            SetupTOTPEvent(
+                                SetupTOTPEvent.EventType.RespondToAuthChallenge(
+                                    eventType.username,
+                                    it.session
+                                )
+                            )
+                        }
+                        else -> {
+                            SetupTOTPEvent(
+                                SetupTOTPEvent.EventType.ThrowAuthError(
+                                    ServiceException(
+                                        message = "An unknown service error has occurred",
+                                        recoverySuggestion = AmplifyException.TODO_RECOVERY_SUGGESTION
+                                    )
+                                )
+                            )
+                        }
+                    }
+                } ?: SetupTOTPEvent(
+                    SetupTOTPEvent.EventType.ThrowAuthError(Exception("Software token verification failed"))
+                )
+            } catch (exception: Exception) {
+                SetupTOTPEvent(
+                    SetupTOTPEvent.EventType.ThrowAuthError(exception)
+                )
+            }
+            logger.verbose("$id Sending event ${evt.type}")
+            dispatcher.send(evt)
+        }
+
+    override fun respondToAuthChallenge(
+        eventType: SetupTOTPEvent.EventType.RespondToAuthChallenge
+    ): Action =
+        Action<AuthEnvironment>("RespondToAuthChallenge") { id, dispatcher ->
+            logger.verbose("$id Starting execution")
+            val evt = try {
+                val challengeResponses = mutableMapOf<String, String>()
```

Comment:
There are a few things missing for the RespondToAuthChallenge (See other RespondToAuthChallenge api calls)

val authParams = mutableMapOf(KEY_USERNAME to event.username, KEY_SRP_A to srpHelper.getPublicA())
                secretHash?.let { authParams[KEY_SECRET_HASH] = it }

                val encodedContextData = getUserContextData(event.username)
                val deviceMetadata = getDeviceMetadata(event.username)
                deviceMetadata?.let { authParams[KEY_DEVICE_KEY] = it.deviceKey }
                val pinpointEndpointId = getPinpointEndpointId()